design/decisions/0005-entity-registration-api.md

```diff
@@ -0,0 +1,266 @@
+  This results in unnecessary duplication of data and makes it error-prone with multiple sources of truth.
+  A central component like `tedge-agent` would be a better fit to store and maintain the entity store.
+* The current entity store also maintains the external id of devices, which is a cloud-specific aspect.
+  When the entity store is moved to the `tedge-agent`, this aspect must be detached and left with the mappers,
+  without the mappers having to maintain a duplicate entity store.
+* When the `tedge-agent` handles the registration of entities, it can't consult with the mappers to make sure that
+  the entities that it accepted locally are valid in the cloud as well.
+  If the local registration succeeds, but the cloud registration fails, the user must get that feedback as well.
+* A mapper might get connected much later, after a lot of local registrations are already processed.
+  Even in such cases, the mapper should be able to fetch all the entities that were registered till then,
+  and replicate that in the cloud.
+  Even in that case, any failed registrations must be conveyed to the user asynchronously.
+
+## Solution Proposal
+
+### Proposal 1: REST APIs for registration
+
+* The `tedge-agent` maintains the entity store and provides REST APIs for registration, and querying.
+  The mappers and other components can query entity data from it when needed.
+* On every successful registration, `tedge-agent` publishes those registration messages to their respective topics,
+  so that other components likes mappers subscribed to those topics are notified of those new entities.
+* The `tedge-agent` can persist the entity store on disk, as it is done today, for faster recovery on restart.
+  It also helps in differentiating already seen entities from newly published ones.
+* The mappers no longer maintains the entire entity store but only maintains a simple mapping
+  between topic ids and their cloud external ids, so that mapped messages can be published to the same.
+* The mappers can rebuild their external id map on every restart, or even persist the same on disk for faster recovery.
+* When the registration of a local entity fails in the cloud for reasons like conflicting external ids,
+  the corresponding error message from the cloud is forwarded to `te/errors` topic so that the user is aware.
+
+#### Create new entity
+
+**Endpoint**
+
+```
+POST /v1/entity
+```
+
+**Payload**
+
+```json
+{
+    "@topic-id": "device/child01//",
+    "@type": "child-device",
+    "@id": "child01",
+    "name": "child01",
+    "extra-fragment": {
+        "extra-key": "extra-value"
+    }
+}
+```
+
+**Responses**
+
+* 201: Created
+  ```json
+  {
+      "@topic-id": "device/child01//"
+  }
+  ```
+* 409: Conflict
+  ```json
+  {
+      "error": "Entity with topic-id: 'device/child01//' already exists."
+  }
+  ```
+
+#### Update existing entity
+
+**Endpoint**
+
+```
+PUT /v1/entity/{topic-id}
```

Comment:
> PUT /v1/entity/{topic-id}

Technically, this should be a `POST` not a `PUT`. Indeed, looking at this example and the following for `GET`, it seems that the intent is to inject new fragments into an existing json resource. A `PUT` implies a full replacement of the resource content.

Reply:
Used `PATCH` instead of `POST` which seemed to be the convention for partial updates to a resource. Kept `PUT` as well for a complete update.

Reply:
Resolved by 34a1e54269dcfaf12b3af4fb99f159b55206825a